File: Exp 9 0_1 Knapsack Problem using Dynamic Programming Approach/0 by 1 Knapsack using dynamic programming.cpp
#include <iostream>
using namespace std;

const int MAXN = 10;   // max number of items
const int MAXW = 50;   // max capacity

int knapsack(int n, int W, int wt[], int val[]) {
	int dp[MAXN + 1][MAXW + 1];

	// Build table dp[][] bottom up
	for (int i = 0; i <= n; i++) {
			for (int w = 0; w <= W; w++) {
					if (i == 0 || w == 0)
							dp[i][w] = 0;
					else if (wt[i - 1] <= w)
							dp[i][w] = (val[i - 1] + dp[i - 1][w - wt[i - 1]] > dp[i - 1][w])
									? (val[i - 1] + dp[i - 1][w - wt[i - 1]])
									: dp[i - 1][w];
					else
							dp[i][w] = dp[i - 1][w];
			}
	}

	return dp[n][W];
}

int main() {
	int n = 4;
	int W = 7;
	int wt[MAXN] = {1, 3, 4, 5};
	int val[MAXN] = {1, 4, 5, 7};

	cout << "Items (value, weight):\n";
	for (int i = 0; i < n; i++) {
			cout << "Item " << (i + 1) << ": value=" << val[i]
			<< " weight=" << wt[i] << "\n";
	}
	cout << "Knapsack capacity = " << W << "\n";

	int maxValue = knapsack(n, W, wt, val);
	cout << "\nMaximum value in knapsack = " << maxValue << "\n";

	return 0;
}

File: Exp 9 0_1 Knapsack Problem using Dynamic Programming Approach/0 by 1 Knapsack using dynamic programming.py
MAXN = 10   # max number of items
MAXW = 50   # max capacity


def knapsack(n, W, wt, val):
    """0/1 Knapsack using dynamic programming"""
    dp = [[0] * (W + 1) for _ in range(MAXN + 1)]
    
    # Build table dp[][] bottom up
    for i in range(n + 1):
        for w in range(W + 1):
            if i == 0 or w == 0:
                dp[i][w] = 0
            elif wt[i - 1] <= w:
                dp[i][w] = max(val[i - 1] + dp[i - 1][w - wt[i - 1]], dp[i - 1][w])
            else:
                dp[i][w] = dp[i - 1][w]
    
    return dp[n][W]


def main():
    n = 4
    W = 7
    wt = [1, 3, 4, 5] + [0] * (MAXN - 4)
    val = [1, 4, 5, 7] + [0] * (MAXN - 4)
    
    print("Items (value, weight):")
    for i in range(n):
        print(f"Item {i + 1}: value={val[i]} weight={wt[i]}")
    print(f"Knapsack capacity = {W}")
    
    max_value = knapsack(n, W, wt, val)
    print(f"\nMaximum value in knapsack = {max_value}")


if __name__ == "__main__":
    main()
